File: tests/python/test_chat_tai_dan.py
"""Khung chat tải dần: mở hội thoại chỉ kéo phần đuôi, cuộn lên mới lấy khúc cũ hơn.

    python tests/run.py chat_tai_dan

Chủ repo báo 17/09: vào một cuộc chat siêu dài thì màn dựng cả nghìn bong bóng một lượt,
khựng vài giây rồi hay đứng lưng chừng thay vì rơi xuống câu trả lời gần nhất.

Hai chỗ dễ làm sai, và test này canh đúng hai chỗ đó:

1. `limit` PHẢI không bắt buộc. Rất nhiều đường khác trong server vẫn gọi get_messages để
   dựng ngữ cảnh cho engine; đổi mặc định thành "chỉ 30 tin cuối" là im lặng cắt trí nhớ
   của Javis giữa cuộc mà không ai thấy lỗi nào.

2. Con trỏ phải là CẶP (ts, id) chứ không chỉ mỗi id. Thứ tự hiển thị là `ORDER BY ts, id`,
   mà tin ghi bù (bot, việc nền) có id lớn trong khi mốc giờ nhỏ. Con trỏ chỉ có id sẽ nhảy
   cóc qua vài tin hoặc trả lại tin đã hiện - cả hai đều lặng thinh.
"""
from _paths import ROOT, SERVER  # noqa: E402,F401
import os
import sys
import tempfile

os.environ.setdefault("JAVIS_STATE_DIR", tempfile.mkdtemp(prefix="javis-taidan-"))

from fastapi.testclient import TestClient  # noqa: E402
import main  # noqa: E402
from sessions import get_store  # noqa: E402

fails = []


def check(name, cond, them=""):
    print(("ok   " if cond else "FAIL ") + name + (("  [" + str(them) + "]") if not cond and them else ""))
    if not cond:
        fails.append(name)


BRAIN = tempfile.mkdtemp(prefix="javis-brain-taidan-")
st = get_store()
SID = "cuoc-sieu-dai"
st.get_or_create(SID, brain=BRAIN, engine="cli", model="x")
for i in range(100):
    st.append_message(SID, "user" if i % 2 == 0 else "assistant", f"tin so {i}")

c = TestClient(main.app, base_url="http://127.0.0.1")


def chu(ds):
    return [m["content"] for m in ds]


# ---- 1. Khuôn cũ không đổi ----------------------------------------------------
d = c.get(f"/sessions/{SID}").json()
check("không truyền limit thì vẫn trả CẢ hội thoại", len(d.get("messages", [])) == 100)
check("thiếu limit thì không kèm has_more (giữ nguyên khuôn cũ)", "has_more" not in d)
check("get_messages của kho phiên vẫn trả đủ", len(st.get_messages(SID)) == 100)

# ---- 2. Mở hội thoại: chỉ phần ĐUÔI ------------------------------------------
d = c.get(f"/sessions/{SID}?limit=30").json()
ds = d.get("messages", [])
check("limit=30 trả đúng 30 tin", len(ds) == 30)
check("và đó là 30 tin CUỐI", chu(ds)[0] == "tin so 70" and chu(ds)[-1] == "tin so 99")
check("xếp xuôi (cũ trước) để dựng bong bóng đúng thứ tự",
      [m["ts"] for m in ds] == sorted(m["ts"] for m in ds))
check("báo còn tin phía trên", d.get("has_more") is True)
check("kèm tổng số tin của cả cuộc", d.get("total") == 100)

# ---- 3. Cuộn lên: khúc cũ hơn -------------------------------------------------
dau = ds[0]
d2 = c.get(f"/sessions/{SID}/messages",
           params={"limit": 30, "before_ts": dau["ts"], "before_id": dau["id"]}).json()
ds2 = d2.get("messages", [])
check("khúc kế trả đúng 30 tin", len(ds2) == 30)
check("nối LIỀN ngay trước khúc trước, không sót không lặp",
      chu(ds2)[0] == "tin so 40" and chu(ds2)[-1] == "tin so 69")
check("vẫn còn tin phía trên", d2.get("has_more") is True)

# Kéo cho tới hết để chắc chắn con trỏ không đứng yên tại chỗ (vòng lặp vô hạn) và không bỏ sót.
gom, truoc, vong = [], ds2[0], 0
while vong < 20:
    vong += 1
    r = c.get(f"/sessions/{SID}/messages",
              params={"limit": 30, "before_ts": truoc["ts"], "before_id": truoc["id"]}).json()
    lo = r.get("messages", [])
    if not lo:
        break
    gom = lo + gom
    truoc = lo[0]
    if not r.get("has_more"):
        break
check("kéo tiếp tới đầu cuộc thì gom đủ 40 tin còn lại", len(gom) == 40, len(gom))
check("và đúng từ tin đầu tiên", chu(gom)[0] == "tin so 0" if gom else False)
check("vòng kéo dừng lại chứ không quay mãi", vong < 20, vong)

# ---- 4. Mốc giờ KHÔNG tăng đều -------------------------------------------------
# Tin ghi bù có id lớn mà ts nhỏ. Con trỏ chỉ dựa vào id sẽ trả lại chính tin đang hiện.
SID2 = "cuoc-lech-gio"
st.get_or_create(SID2, brain=BRAIN, engine="cli", model="x")
st.append_message(SID2, "user", "A")
st.append_message(SID2, "assistant", "B")
with st._lock:  # ghi thẳng một tin có id LỚN NHẤT nhưng mốc giờ SỚM NHẤT
    st._conn.execute("INSERT INTO messages (session_id, role, content, ts) VALUES (?,?,?,?)",
                     (SID2, "user", "Z-som-nhat", 1.0))
    st._conn.commit()
d3 = c.get(f"/sessions/{SID2}?limit=2").json()
ds3 = d3.get("messages", [])
check("tin mốc giờ sớm nhất không lọt vào khúc cuối dù id lớn nhất",
      chu(ds3) == ["A", "B"], chu(ds3))
d4 = c.get(f"/sessions/{SID2}/messages",
           params={"limit": 5, "before_ts": ds3[0]["ts"], "before_id": ds3[0]["id"]}).json()
check("cuộn lên lấy đúng tin ghi bù, không trả lại tin đang hiện",
      chu(d4.get("messages", [])) == ["Z-som-nhat"], chu(d4.get("messages", [])))
check("tới đầu cuộc thì has_more tắt", d4.get("has_more") is False)

# ---- 5. Ca biên ---------------------------------------------------------------
check("phiên không tồn tại trả 404", c.get("/sessions/khong-co-that/messages").status_code == 404)
st.get_or_create("cuoc-rong", brain=BRAIN, engine="cli", model="x")
d5 = c.get("/sessions/cuoc-rong?limit=30").json()
check("cuộc chưa có tin nào: rỗng và không còn gì phía trên",
      d5.get("messages") == [] and d5.get("has_more") is False and d5.get("total") == 0)
check("thiếu con trỏ thì rơi về khúc cuối, không nổ",
      chu(c.get(f"/sessions/{SID}/messages", params={"limit": 3}).json()["messages"])
      == ["tin so 97", "tin so 98", "tin so 99"])

print("")
if fails:
    print(f"ĐỎ {len(fails)} mục")
    sys.exit(1)
print("Tất cả xanh.")

File: tests/js/test_chat_tai_dan.js
/* Khung chat tải dần + phân trang trang Kỹ năng.

       node tests/js/test_chat_tai_dan.js

   Hai việc đi chung một file vì cùng một bài toán: danh sách dài thì đừng đổ hết ra một lượt.

   Phần khung chat có ba cái bẫy, mỗi cái đều IM LẶNG khi làm sai:

   1. Chèn ngược mà quên chặn cuộn: appendUserMessage/appendJavisMessage đều gọi scrollBottom
      theo thói quen, nên chèn xong là màn quăng thẳng xuống đáy - người đọc đang xem giữa
      cuộc bị đá đi mà không hiểu vì sao.

   2. Giữ chỗ cuộn phải đo theo ĐÁY khung (scrollHeight - scrollTop), không phải theo
      scrollTop. Phần chèn nằm phía trên nên scrollTop cũ trỏ vào một chỗ khác hẳn.

   3. Mồi tải tin chỉ được chèn KHI CÒN tin cũ. `.transcript:empty::after` là câu mời "Nói
      hoặc gõ để bắt đầu"; một node con thường trực là câu đó biến mất lặng lẽ - đúng cái bẫy
      đã ghi sẵn trong app.js cho #newMsgBtn và trong chat-marks.js.

   Phần trang Kỹ năng canh chuyện chép khối: repo đã có sẵn pager() dùng chung cho nhật ký
   loop, trang Tự học và trang Tiết kiệm. Viết bản thứ hai cho trang Kỹ năng là hai bản trôi
   lệch nhau ngay lần sửa đầu tiên. */
const fs = require("fs");
const path = require("path");

const ROOT = path.join(__dirname, "..", "..");
const D = (f) => fs.readFileSync(path.join(ROOT, "dashboard", f), "utf8");
const APP = D("app.js");
const CONSOLE = D("console.js");
const STUDIO = D("studio.js");
const CSS = D("style.css");
const VI = JSON.parse(D("i18n/vi.json"));
const EN = JSON.parse(D("i18n/en.json"));

const fails = [];
const check = (name, cond) => { console.log((cond ? "ok   " : "FAIL ") + name); if (!cond) fails.push(name); };

// ============================================================
// A. Khung chat chỉ dựng phần đuôi
// ============================================================
check("mở hội thoại cũ kèm limit, không kéo cả cuộc về",
  /\/sessions\/\$\{encodeURIComponent\(id\)\}\?limit=\$\{TIN_MOI_LUOT\}/.test(APP));
check("có hằng số số tin mỗi lượt", /const TIN_MOI_LUOT = \d+;/.test(APP));
check("vẫn rơi xuống đáy sau khi dựng xong",
  /scrollBottom\(true\);\s*\n\s*ghimDay\(ticket\);\s*\n\s*notifySessions\(\);/.test(APP));
// Đặt scrollTop MỘT LẦN là không đủ: thanh mốc hội thoại chèn vào rồi bật lề phải làm bong
// bóng xuống dòng, ảnh tải xong mới đẩy chiều cao ra. Đo thật trên cuộc 120 tin: mở xong
// đứng cách đáy 174px, câu trả lời gần nhất bị cắt. Nên phải ghim lại trong một quãng ngắn.
check("ghim lại đáy qua vài nhịp vẽ, vì khung còn cao thêm sau lượt mở",
  /function ghimDay\(ticket, ms\)/.test(APP) && /requestAnimationFrame\(nhip\)/.test(APP));
check("nhả ghim ngay khi người dùng tự cuộn",
  /addEventListener\("wheel", nhaTay/.test(APP) && /addEventListener\("touchmove", nhaTay/.test(APP));
check("và nhả luôn khi đổi sang hội thoại khác", /if \(thoi \|\| ticket !== sessionOpenSeq\)/.test(APP));
// Con trỏ lấy từ tin ĐẦU của khúc đang hiện, tức tin già nhất trên màn.
check("giữ con trỏ (ts, id) của tin già nhất đang hiện",
  /ts: ds\[0\]\.ts, id: ds\[0\]\.id, con: !!sess\.has_more/.test(APP));

// ============================================================
// B. Cuộn lên tải tiếp
// ============================================================
check("có hàm tải khúc cũ hơn", /async function taiTinCu\(\)/.test(APP));
check("gọi đúng đường /messages với cặp con trỏ",
  APP.indexOf("/messages?limit=") !== -1 &&
  APP.indexOf("before_ts=") !== -1 && APP.indexOf("before_id=") !== -1);
check("dùng IntersectionObserver để tự tải khi cuộn tới đầu khung",
  /new IntersectionObserver\(/.test(APP) && /root: chatArea/.test(APP));
check("còn nút bấm tay khi trình duyệt không có IntersectionObserver",
  /typeof IntersectionObserver !== "function"/.test(APP));
// Bẫy 1.
check("có cờ chặn cuộn trong lúc chèn ngược", /let _neoChenCu = null, _dangChenCu = false;/.test(APP));
check("scrollBottom thoát sớm khi đang chèn ngược",
  /function scrollBottom\(force\) \{[\s\S]{0,400}?if \(_dangChenCu\) return;/.test(APP));
check("chatAppend chèn trước cái neo khi có neo",
  /chatArea\.insertBefore\(el, _neoChenCu \|\| newMsgBtn\)/.test(APP));
// Bẫy 2.
check("giữ chỗ cuộn đo theo ĐÁY khung, không theo scrollTop",
  /const cachDay = chatArea\.scrollHeight - chatArea\.scrollTop;/.test(APP) &&
  /chatArea\.scrollTop = chatArea\.scrollHeight - cachDay;/.test(APP));
check("đặt lại chỗ cuộn SAU khi dựng lại mồi, không phải trước",
  APP.indexOf("datMoiTinCu();\n    chatArea.scrollTop = chatArea.scrollHeight - cachDay;") !== -1);
// Đổi phiên giữa chừng: khúc vừa về là của cuộc khác.
check("vứt khúc về muộn khi đã đổi phiên",
  /_tinCu !== st \|\| st\.sid !== savedSessionId/.test(APP));
check("khoá không cho hai lượt tải chồng nhau", /if \(!st \|\| !st\.con \|\| st\.dangTai\) return;/.test(APP));
// Bẫy 3.
check("chỉ chèn mồi khi CÒN tin cũ", /if \(!_tinCu \|\| !_tinCu\.con\) return;/.test(APP));
check("mở hội thoại mới thì dọn mồi và con trỏ",
  /_tinCu = null; goMoiTinCu\(\);/.test(APP));
// Neo theo bong bóng đầu tiên chứ không theo firstChild: đầu khung còn có đồ nội thất dán
// sẵn (thanh mốc hội thoại) tính là mình vẫn đứng đầu, chen lên trước là đẩy nó khỏi chỗ.
check("mồi neo theo bong bóng đầu tiên, không phải firstChild",
  /function dauKhungChat\(\)/.test(APP) &&
  /return chatArea\.querySelector\("\.msg"\);/.test(APP) &&
  !/dauKhungChat[\s\S]{0,200}chatArea\.firstChild/.test(APP));

// ============================================================
// C. Sống sót qua F5
// ============================================================
check("con trỏ tin cũ được lưu xuống localStorage", /tinCu: \(_tinCu && convo\.length <= 200\)/.test(APP));
check("khôi phục phiên thì dựng lại mồi", /_tinCu = \(s\.tinCu && savedSessionId\)/.test(APP));

// ============================================================
// D. Chữ hiển thị + CSS
// ============================================================
["app.older_load", "app.older_loading"].forEach(k => {
  check("có chuỗi " + k + " ở cả vi và en", !!VI[k] && !!EN[k]);
});
check("chuỗi tiếng Việt có dấu", /[ăâêôơưđáàảãạ]/i.test(VI["app.older_load"]));
check("có CSS cho mồi tải tin cũ", CSS.indexOf(".older-seed") !== -1 && CSS.indexOf(".older-btn") !== -1);

// ============================================================
// E. Phân trang trang Kỹ năng - dùng LẠI pager() sẵn có
// ============================================================
check("console.js xuất pager ra cho module khác", /window\.JavisPager = pager;/.test(CONSOLE));
check("pager nhận được cả Node lẫn chuỗi HTML", /if \(ruot && ruot\.nodeType\)/.test(CONSOLE));
check("trang Kỹ năng gọi pager dùng chung, không tự viết bản thứ hai",
  /window\.JavisPager\(box, list, SK_MOI_TRANG, veTrang\)/.test(STUDIO));
check("có hằng số số skill mỗi trang", /const SK_MOI_TRANG = \d+;/.test(STUDIO));
check("thẻ skill tách thành hàm riêng để mỗi trang dựng lại được",
  /function theSkill\(s\) \{/.test(STUDIO) && /phan\.forEach\(s => fr\.appendChild\(theSkill\(s\)\)\)/.test(STUDIO));
// Chọn tất cả phải gom CẢ danh sách đang lọc, không phải mỗi trang đang hiện - đúng bài học
// đã ghi trong test_chon_skill_va_phan_trang.js: lọc rồi đọc trạng thái từ DOM là mất tick.
check("Chọn tất cả vẫn gom cả danh sách đang lọc, không chỉ trang đang hiện",
  /_skFiltered\(\)\.filter\(s => !s\.system\)\.map\(s => s\.slug\)/.test(STUDIO));
check("tick phục hồi từ Set khi lật trang", /el\.checked = _sel\[kind\]\.has\(slug\);/.test(STUDIO));
check("hàng phân trang trong danh sách skill được căn giữa",
  /\.sk2-list \.jv-pager\{justify-content:center\}/.test(STUDIO));

console.log("");
if (fails.length) { console.log("ĐỎ " + fails.length + " mục"); process.exit(1); }
console.log("Tất cả xanh.");
